File: ns-3.23/scratch/bigcsma.cc
#include <iostream>

#include "ns3/core-module.h"
#include "ns3/point-to-point-module.h"
#include "ns3/network-module.h"
#include "ns3/applications-module.h"
#include "ns3/wifi-module.h"
#include "ns3/mobility-module.h"
#include "ns3/csma-module.h"
#include "ns3/internet-module.h"

using namespace ns3;

float average (int val1, int val2) {
  return float (val1 + val2) / 2;
}

int main (int argc, char *argv[]) {
  // TODO: create an array of values
  // TODO: create an array of Boolean flags
  std::cout << "hi there" << std::endl;
  std::cout << "(5 + 6) / 2 = " << average(5, 6) << std::endl;

  uint32_t nCsma = 100;
  CommandLine cmd;
  cmd.AddValue ("nCsma", "Number of CSMA nodes", nCsma);
  cmd.Parse (argc, argv);

  LogComponentEnable ("UdpEchoClientApplication", LOG_LEVEL_INFO);
  LogComponentEnable ("UdpEchoServerApplication", LOG_LEVEL_INFO);

  NodeContainer csmaNodes;
  csmaNodes.Create (nCsma);

  CsmaHelper csma;
  csma.SetChannelAttribute ("DataRate", StringValue ("100Mbps"));
  csma.SetChannelAttribute ("Delay", TimeValue (NanoSeconds (6560)));

  NetDeviceContainer csmaDevices;
  csmaDevices = csma.Install (csmaNodes);

  InternetStackHelper stack;
  stack.Install (csmaNodes);

  Ipv4AddressHelper address;
  address.SetBase ("10.1.1.0", "255.255.255.0");
  Ipv4InterfaceContainer csmaInterfaces;
  csmaInterfaces = address.Assign (csmaDevices);

  UdpEchoServerHelper echoServer (9);
  ApplicationContainer serverApps = echoServer.Install (csmaNodes.Get (nCsma - 1));
  serverApps.Start (Seconds (0.0));
  //serverApps.Stop (Seconds (nCsma));

  for (uint32_t i = 0; i < nCsma-1; i++) {
    UdpEchoClientHelper echoClient (csmaInterfaces.GetAddress (nCsma - 1), 9);
    echoClient.SetAttribute ("MaxPackets", UintegerValue (1));
    echoClient.SetAttribute ("Interval", TimeValue(Seconds (1.0)));
    echoClient.SetAttribute ("PacketSize", UintegerValue (1024));
    ApplicationContainer clientApps = echoClient.Install (csmaNodes.Get (i));
    clientApps.Start (Seconds (1));
    //clientApps.Stop (Seconds (10));
  }

  Ipv4GlobalRoutingHelper::PopulateRoutingTables ();

  Simulator::Run ();
  Simulator::Destroy ();
  return 0;
}

File: ns-3.23/scratch/bigcsma.py
import ns.applications
import ns.core
import ns.csma
import ns.internet
import ns.network

nCsma = 3
csmaDataRate = "100Mbps"
csmaDelay = "2ms"

IP_ADDRESS_BASE = "10.1.0.0"
IP_ADDRESS_MASK = "255.255.0.0"

ns.core.LogComponentEnable('UdpClient', ns.core.LOG_LEVEL_INFO)
#ns.core.LogComponentEnable('UdpEchoServerApplication', ns.core.LOG_LEVEL_INFO)

nodes = ns.network.NodeContainer()
nodes.Create(nCsma)

csma = ns.csma.CsmaHelper()
csma.SetChannelAttribute("DataRate", ns.core.StringValue("100Mbps"))
csma.SetChannelAttribute("Delay", ns.core.StringValue("2ms"))

devices = csma.Install(nodes)

stack = ns.internet.InternetStackHelper()
stack.Install(nodes)

address = ns.internet.Ipv4AddressHelper()
address.SetBase(ns.network.Ipv4Address(IP_ADDRESS_BASE), ns.network.Ipv4Mask(IP_ADDRESS_MASK))

interfaces = address.Assign(devices)

print "creating server apps"

server = ns.applications.UdpEchoServerHelper(9)
#for i in xrange(nCsma):
serverApps = server.Install(nodes)

serverApps.Start(ns.core.Seconds(0))


print "creating client apps"
clientApps = ns.network.ApplicationContainer()
for i in xrange(nCsma):
    echoClient = ns.applications.UdpClientHelper(interfaces.GetAddress(i), 9)
    echoClient.SetAttribute("MaxPackets", ns.core.UintegerValue(2))
    echoClient.SetAttribute("Interval", ns.core.TimeValue(ns.core.Seconds(0.0)))
    echoClient.SetAttribute("PacketSize", ns.core.UintegerValue(1024))
    node_container = ns.network.NodeContainer()
    node_container.Add(nodes.Get(i))
    clientApps.Add(echoClient.Install(node_container))
    #print clientApps
print dir(clientApps.Get(0))

print "starting client apps"
clientApps.Start(ns.core.Seconds(0))
clientApps.Stop(ns.core.Seconds(100))

def scheduled_callback():
    print "callback at %.3ss" % (ns.core.Simulator.Now().GetMilliSeconds() / 1000.0)

ns.core.Simulator.Schedule(ns.core.Seconds(2), scheduled_callback)

ns.core.Simulator.Run()
ns.core.Simulator.Destroy()
